File: 2023/day08/php/solution.php
<?php
/**
 * Advent of Code 2023 - Day 8: Haunted Wasteland
 *
 * Navigate a network following L/R instructions.
 * Part 1: Go from AAA to ZZZ
 * Part 2: Start from all nodes ending in 'A' simultaneously,
 *         find when all reach nodes ending in 'Z'
 */

/**
 * Parse the input into instructions and network.
 */
function parseInput(string $text): array {
    $lines = explode("\n", trim($text));
    $instructions = $lines[0];

    $network = [];
    $lineCount = count($lines);
    for ($i = 2; $i < $lineCount; $i++) {
        $line = trim($lines[$i]);
        if (empty($line)) {
            continue;
        }
        // Parse: AAA = (BBB, CCC)
        preg_match('/^(\w+) = \((\w+), (\w+)\)$/', $line, $matches);
        $network[$matches[1]] = [$matches[2], $matches[3]];
    }

    return [$instructions, $network];
}

/**
 * Part 1: Navigate from AAA to ZZZ following L/R instructions.
 */
function part1(string $instructions, array $network): int {
    $current = 'AAA';
    $steps = 0;
    $instructionLen = strlen($instructions);

    while ($current !== 'ZZZ') {
        $instruction = $instructions[$steps % $instructionLen];
        $current = $instruction === 'L' ? $network[$current][0] : $network[$current][1];
        $steps++;
    }

    return $steps;
}

/**
 * Part 2: Navigate all nodes ending in A simultaneously to nodes ending in Z.
 */
function part2(string $instructions, array $network): string {
    // Find all starting nodes (ending in A)
    $startNodes = array_filter(array_keys($network), fn($node) => str_ends_with($node, 'A'));

    $instructionLen = strlen($instructions);
    $cycleLengths = [];

    // For each starting node, find how many steps to reach a Z node
    foreach ($startNodes as $node) {
        $current = $node;
        $steps = 0;

        while (!str_ends_with($current, 'Z')) {
            $instruction = $instructions[$steps % $instructionLen];
            $current = $instruction === 'L' ? $network[$current][0] : $network[$current][1];
            $steps++;
        }

        $cycleLengths[] = $steps;
    }

    // Find LCM of all cycle lengths using GMP
    $result = array_reduce(
        $cycleLengths,
        fn($carry, $item) => gmp_lcm($carry, $item),
        gmp_init(1)
    );

    return gmp_strval($result);
}

// Main execution
$inputPath = __DIR__ . '/../input.txt';
$text = file_get_contents($inputPath);

[$instructions, $network] = parseInput($text);

echo "Part 1: " . part1($instructions, $network) . "\n";
echo "Part 2: " . part2($instructions, $network) . "\n";

File: 2023/day08/clojure/solution.clj
#!/usr/bin/env bb
;; Advent of Code 2023 - Day 8: Haunted Wasteland

(require '[clojure.string :as str]
         '[clojure.java.io :as io])

(defn parse-input
  "Parse the input into instructions and network map."
  [text]
  (let [lines (str/split-lines (str/trim text))
        instructions (first lines)
        network (->> (drop 2 lines)
                     (remove str/blank?)
                     (map (fn [line]
                            (let [[node rest] (str/split line #" = ")
                                  [left right] (-> rest
                                                   (subs 1 (dec (count rest)))
                                                   (str/split #", "))]
                              [node [left right]])))
                     (into {}))]
    [instructions network]))

(defn follow-instructions
  "Follow instructions from start until end condition is met."
  [instructions network start end-pred]
  (loop [current start
         steps 0
         instrs (cycle instructions)]
    (if (end-pred current)
      steps
      (let [[left right] (network current)
            next-node (if (= (first instrs) \L) left right)]
        (recur next-node (inc steps) (rest instrs))))))

(defn part1
  "Navigate from AAA to ZZZ following L/R instructions."
  [instructions network]
  (follow-instructions instructions network "AAA" #(= % "ZZZ")))

(defn gcd
  "Greatest common divisor."
  [a b]
  (if (zero? b)
    a
    (recur b (mod a b))))

(defn lcm
  "Least common multiple."
  [a b]
  (/ (* a b) (gcd a b)))

(defn part2
  "Navigate all nodes ending in A simultaneously to nodes ending in Z."
  [instructions network]
  (let [start-nodes (filter #(str/ends-with? % "A") (keys network))
        cycle-lengths (map #(follow-instructions instructions network % (fn [s] (str/ends-with? s "Z")))
                           start-nodes)]
    (reduce lcm 1 cycle-lengths)))

(defn -main []
  (let [input-file (io/file (System/getProperty "user.dir") ".." "input.txt")
        text (slurp input-file)
        [instructions network] (parse-input text)]
    (println "Part 1:" (part1 instructions network))
    (println "Part 2:" (part2 instructions network))))

(-main)
